From 2452bbb6f6f094d5621c056c7dd5e25a4168ac49 Mon Sep 17 00:00:00 2001
From: Anirudha Sarangi <anirudh@xilinx.com>
Date: Fri, 18 Apr 2014 18:45:37 +0530
Subject: [STANDALONE] bsp: Support for msrset and msrclr instructions

This patch fixes the CR#764881. It adds support for msrset and
msrclr instructions. It renames some of the *.s files into *.S files
so that the defines in xparameters.h can be used in them.

Signed-off-by: Anirudha Sarangi <anirudh@xilinx.com>
---
 lib/bsp/standalone_v4_1/data/standalone.tcl        |    2 +-
 lib/bsp/standalone_v4_1/src/changelog.txt          |    3 ++
 .../src/microblaze/microblaze_disable_dcache.S     |   29 ++++++++++++++-----
 ...xceptions.s => microblaze_disable_exceptions.S} |   15 +++++++---
 ...isable_icache.s => microblaze_disable_icache.S} |   11 +++++++-
 ...nterrupts.s => microblaze_disable_interrupts.S} |    9 +++++-
 ..._enable_dcache.s => microblaze_enable_dcache.S} |   10 ++++++-
 ...exceptions.s => microblaze_enable_exceptions.S} |   17 ++++++++---
 ..._enable_icache.s => microblaze_enable_icache.S} |   11 ++++++-
 ...interrupts.s => microblaze_enable_interrupts.S} |    9 +++++-
 10 files changed, 92 insertions(+), 24 deletions(-)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_disable_exceptions.s => microblaze_disable_exceptions.S} (86%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_disable_icache.s => microblaze_disable_icache.S} (90%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_disable_interrupts.s => microblaze_disable_interrupts.S} (90%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_enable_dcache.s => microblaze_enable_dcache.S} (89%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_enable_exceptions.s => microblaze_enable_exceptions.S} (80%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_enable_icache.s => microblaze_enable_icache.S} (88%)
 rename lib/bsp/standalone_v4_1/src/microblaze/{microblaze_enable_interrupts.s => microblaze_enable_interrupts.S} (90%)

diff --git a/lib/bsp/standalone_v4_1/data/standalone.tcl b/lib/bsp/standalone_v4_1/data/standalone.tcl
index 644175d..0cb0f69 100755
--- a/lib/bsp/standalone_v4_1/data/standalone.tcl
+++ b/lib/bsp/standalone_v4_1/data/standalone.tcl
@@ -126,7 +126,7 @@ proc generate {os_handle} {
     set makeconfig [open "./src/config.make" w]
 #    print_generated_header_tcl $makeconfig "Configuration parameters for Standalone Makefile"
     if { $proctype == "microblaze" } {
-        puts $makeconfig "LIBSOURCES = *.c *.s *.S"
+        puts $makeconfig "LIBSOURCES = *.c *.S"
         puts $makeconfig "PROFILE_ARCH_OBJS = profile_mcount_mb.o"
     } elseif { $proctype == "ps7_cortexa9" } {
         if {[string compare -nocase $compiler "armcc"] == 0} {
diff --git a/lib/bsp/standalone_v4_1/src/changelog.txt b/lib/bsp/standalone_v4_1/src/changelog.txt
index 7a1c692..69f5fc3 100755
--- a/lib/bsp/standalone_v4_1/src/changelog.txt
+++ b/lib/bsp/standalone_v4_1/src/changelog.txt
@@ -163,4 +163,7 @@
  *		       microblaze_invalidate_cache_ext and microblaze_flush_cache_ext in
  *		       mb_interface.h
  *
+ * 4.1. asa  04/18/14  Fix for CR#764881. Added support for msrset and msrclr. Renamed
+ *		       some of the *.s files inMB BSP source to *.S.
+ *
  *****************************************************************************************/
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_dcache.S b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_dcache.S
index bb816c5..d1f1296 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_dcache.S
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_dcache.S
@@ -31,23 +31,36 @@
 	.ent	microblaze_disable_dcache
 	.align	2
 microblaze_disable_dcache:
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
 
-        addik   r1, r1, -28
+#if XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0
+    swi     r15, r1, -4
+    brlid   r15, microblaze_flush_dcache /* microblaze_flush_dcache does not use r1*/
+    nop
+    lwi     r15, r1, 4
+#endif /* XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0 */
+	rtsd	r15, 8
+	msrclr r0, 128
+
+#else /* XPAR_MICROBLAZE_USE_MSR_INSTR == 1 */
+
+    addik   r1, r1, -28
 
 #if XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0
-        swi     r15, r1, 0
-        brlid   r15, microblaze_flush_dcache
-        nop
-#endif
-    	mfs	r11, rmsr
+    swi     r15, r1, 0
+    brlid   r15, microblaze_flush_dcache
+    nop
+#endif /* XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0 */
+    mfs	r11, rmsr
 	andi	r11, r11, ~128
 	mts	rmsr, r11
 
 #if XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0   
-        lwi     r15, r1, 0
-#endif
+    lwi     r15, r1, 0
+#endif /* XPAR_MICROBLAZE_DCACHE_USE_WRITEBACK != 0 */
 
 	rtsd	r15, 8
 	addi	r1, r1, 28
 
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
         .end	microblaze_disable_dcache
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.S
similarity index 86%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.S
index feee265..30ba32d 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_exceptions.S
@@ -21,16 +21,23 @@
 #
 ####################################################################
 
-.section .text
-.globl	microblaze_disable_exceptions
-.ent	microblaze_disable_exceptions
-.align 2
+#include "xparameters.h"
+
+	.text
+	.globl	microblaze_disable_exceptions
+	.ent	microblaze_disable_exceptions
+	.align 2
 microblaze_disable_exceptions:
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd    r15, 8
+        msrclr  r0, 0x100
+#else 
         mfs     r4, rmsr;
         andi    r4, r4, ~(0x100);                       /* Turn OFF the EE bit */
         mts     rmsr, r4;
         rtsd    r15, 8;
         nop;        
+#endif
 .end microblaze_disable_exceptions
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.S
similarity index 90%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.S
index 879df67..8665f9c 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_icache.S
@@ -25,12 +25,20 @@
 # $Id: microblaze_disable_icache.s,v 1.1.2.1 2011/05/17 04:37:27 sadanan Exp $
 #
 ####################################################################
-	
+
+#include "xparameters.h"
+
 	.text
 	.globl	microblaze_disable_icache
 	.ent	microblaze_disable_icache
 	.align	2
 microblaze_disable_icache:	
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd	r15, 8
+	msrclr  r0, 32
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
+	rtsd	r15, 8
+	msrclr  r0, 32
 	#Make space on stack for a temporary
 	addi	r1, r1, -4
 	#Read the MSR register
@@ -43,6 +51,7 @@ microblaze_disable_icache:
 	rtsd	r15, 8
 	#Update stack in the delay slot
 	addi	r1, r1, 4
+#endif
 	.end	microblaze_disable_icache
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.S
similarity index 90%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.S
index 1956120..d2b71ee 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_disable_interrupts.S
@@ -25,12 +25,18 @@
 # $Id: microblaze_disable_interrupts.s,v 1.1.2.1 2011/05/17 04:37:27 sadanan Exp $
 #
 ####################################################################
-	
+
+#include "xparameters.h"
+
 	.text
 	.globl	microblaze_disable_interrupts
 	.ent	microblaze_disable_interrupts
 	.align	2
 microblaze_disable_interrupts:	
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd	r15, 8
+	msrclr  r0, 2
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	#Make space on stack for a temporary
 	addi	r1, r1, -4
 	#Save register r12
@@ -47,6 +53,7 @@ microblaze_disable_interrupts:
 	rtsd	r15, 8
 	#Update stack in the delay slot
 	addi	r1, r1, 4
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	.end	microblaze_disable_interrupts
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.S
similarity index 89%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.S
index 433364d..d010279 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_dcache.S
@@ -25,12 +25,19 @@
 # $Id: microblaze_enable_dcache.s,v 1.1.2.1 2011/05/17 04:37:27 sadanan Exp $
 #
 ####################################################################
-	
+
+#include "xparameters.h"	
+
 	.text
 	.globl	microblaze_enable_dcache
 	.ent	microblaze_enable_dcache
 	.align	2
 microblaze_enable_dcache:	
+
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd	r15, 8
+	msrset  r0, 128
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	#Make space on stack for a temporary
 	addi	r1, r1, -4
 	#Read the MSR register
@@ -43,6 +50,7 @@ microblaze_enable_dcache:
 	rtsd	r15, 8
 	#Update stack in the delay slot
 	addi	r1, r1, 4
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	.end	microblaze_enable_dcache
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.S
similarity index 80%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.S
index 72dc6fa..fd74376 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_exceptions.S
@@ -21,16 +21,23 @@
 #
 ####################################################################
 
-.section .text
-.globl	microblaze_enable_exceptions
-.ent	microblaze_enable_exceptions
-.align 2
+#include "xparameters.h"
+
+	.text
+	.globl	microblaze_enable_exceptions
+	.ent	microblaze_enable_exceptions
+	.align 2
 microblaze_enable_exceptions:
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd    r15, 8;
+        msrset  r0, 0x100
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
         mfs     r4, rmsr;
         ori     r4, r4, 0x100;                  /* Turn ON the EE bit */
         mts     rmsr, r4;
         rtsd    r15, 8;
-        nop;        
+        nop;       
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 .end microblaze_enable_exceptions
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.S
similarity index 88%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.S
index 379e446..6224a4b 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_icache.S
@@ -25,12 +25,18 @@
 # $Id: microblaze_enable_icache.s,v 1.1.2.1 2011/05/17 04:37:27 sadanan Exp $
 #
 ####################################################################
-	
+
+#include "xparameters.h"
+
 	.text
 	.globl	microblaze_enable_icache
 	.ent	microblaze_enable_icache
 	.align	2
-microblaze_enable_icache:	
+microblaze_enable_icache:
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd	r15, 8
+	msrset  r0, 32
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	#Make space on stack for a temporary
 	addi	r1, r1, -4
 	#Read the MSR register
@@ -43,6 +49,7 @@ microblaze_enable_icache:
 	rtsd	r15, 8
 	#Update stack in the delay slot
 	addi	r1, r1, 4
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	.end	microblaze_enable_icache
 
 	
diff --git a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.s b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.S
similarity index 90%
rename from lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.s
rename to lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.S
index 063feb8..6248d02 100755
--- a/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.s
+++ b/lib/bsp/standalone_v4_1/src/microblaze/microblaze_enable_interrupts.S
@@ -25,12 +25,18 @@
 # $Id: microblaze_enable_interrupts.s,v 1.1.2.1 2011/05/17 04:37:28 sadanan Exp $
 #
 ####################################################################
-	
+
+#include "xparameters.h"	
+
 	.text
 	.globl	microblaze_enable_interrupts
 	.ent	microblaze_enable_interrupts
 	.align	2
 microblaze_enable_interrupts:	
+#if XPAR_MICROBLAZE_USE_MSR_INSTR == 1
+	rtsd	r15, 8
+	msrset  r0, 2
+#else /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	#Make space on stack for a temporary
 	addi	r1, r1, -4
 	#Save register r12
@@ -47,6 +53,7 @@ microblaze_enable_interrupts:
 	rtsd	r15, 8
 	#Update stack in the delay slot
 	addi	r1, r1, 4
+#endif /*XPAR_MICROBLAZE_USE_MSR_INSTR == 1*/
 	.end	microblaze_enable_interrupts
 
 	
-- 
1.7.4

